src/WebSdk/Publish/Tasks/Tasks/MsDeploy/MSDeploy.cs

```diff
@@ -263,13 +264,24 @@ public string ExePath
 #else
                 if (string.IsNullOrEmpty(m_exePath))
                 {
-                    string programFilesX86 = System.Environment.ExpandEnvironmentVariables("%ProgramFiles(x86)%");
-                    string msdeployExePath = System.IO.Path.Combine("IIS", "Microsoft Web Deploy V3");
-                    m_exePath = IO.Path.Combine(programFilesX86, msdeployExePath);
+                    string programFiles = Environment.ExpandEnvironmentVariables("%ProgramFiles(x86)%");
+                    string msdeployExePath = IO.Path.Combine("IIS", "Microsoft Web Deploy V3");
+                    m_exePath = IO.Path.Combine(programFiles, msdeployExePath);
                     if (!IO.File.Exists(IO.Path.Combine(m_exePath, ToolName)))
                     {
-                        string programFiles = System.Environment.ExpandEnvironmentVariables("%ProgramW6432%");
-                        m_exePath = IO.Path.Combine(programFiles, msdeployExePath);
+                        /// On 32-bit Operating Systems, this will return C:\Program Files
+                        /// On 64-bit Operating Systems - regardless of process bitness, this will return C:\Program Files
+                        if (!Environment.Is64BitOperatingSystem || Environment.Is64BitProcess)
+                        {
+                            programFiles = Environment.GetFolderPath(Environment.SpecialFolder.ProgramFiles);
+                            m_exePath = IO.Path.Combine(programFiles, msdeployExePath);
+                        }
+                        else
+                        {
+                            // 32 bit process on a 64 bit OS can't use SpecialFolder.ProgramFiles to get the 64-bit program files folder
```

Comment:
I thought the KnownFolder API might be more reliable than an environment variable, and there is FOLDERID_ProgramFilesX64 dedicated to this folder, but the docs say it is not supported in 32-bit processes. ☹️

Reply:
Went with the existing env variable approach because it can help fix issues like this #25887 where users can set the env variable themselves to bypass the issue.